jax/experimental/callback.py

```diff
@@ -0,0 +1,156 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as onp
+from typing import Any, Callable, Dict, Optional, Sequence, Tuple, Union
+
+import jax.numpy as np
+
+from jax import core
+from jax.core import Trace, Tracer, new_master
+from jax import linear_util as lu
+from jax.util import partial, safe_map, wrap_name
+from jax.interpreters import xla
+from jax.interpreters import partial_eval as pe
+
+import inspect
+from jax.api_util import (wraps, flatten_fun_nokwargs)
+from jax.tree_util import (tree_flatten, tree_unflatten)
+from jax import lax
+
+map = safe_map
+
+### Public
+
+def callback_transform(
+    fun: Callable, callback: Callable, strip_calls=False) -> Callable:
+  _check_callable(fun)
+
+  @wraps(fun)
+  def wrapped_fun(*args):
+    args_flat, in_tree  = tree_flatten(args)
+    f = lu.wrap_init(fun)
+    flat_fun, out_tree = flatten_fun_nokwargs(f, in_tree)
+    out_flat = callback_fun(flat_fun, args_flat, callback, strip_calls)
+    return tree_unflatten(out_tree(), out_flat)
+
+  return wrapped_fun
+
+### Example Transform
+
+def find_by_value(fun: Callable, queries) -> Callable:
+  def find_callback(p, vals, params):
```

Comment:
The name `p` is a bit short, and `prim` may be better. But I leave that up to you! More importantly, instead of relying on the name alone for info, how about we add a type annotation? I think we can write `prim: core.Primitive` for example. (We could also write `vals: Sequence[core.Tracer]` and `params : Dict`, and a return type annotation `-> Union[core.Tracer, Sequence[core.Tracer]]`.) I'm really into these type annotations as docs.

Reply:
Great idea!